services/agit/agit.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package agit
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/notification"
+	"code.gitea.io/gitea/modules/private"
+	pull_service "code.gitea.io/gitea/services/pull"
+)
+
+// ProcRecive handle proc receive work
+func ProcRecive(ctx *context.PrivateContext, opts *private.HookOptions) []private.HockProcReceiveRefResult {
+	// TODO: Add more options?
+	var (
+		topicBranch string
+		title       string
+		description string
+		forcePush   bool
+	)
+
+	results := make([]private.HockProcReceiveRefResult, 0, len(opts.OldCommitIDs))
+	repo := ctx.Repo.Repository
+	gitRepo := ctx.Repo.GitRepo
+	ownerName := ctx.Repo.Repository.OwnerName
+	repoName := ctx.Repo.Repository.Name
+
+	topicBranch = opts.GitPushOptions["topic"]
+	_, forcePush = opts.GitPushOptions["force-push"]
+
+	for i := range opts.OldCommitIDs {
+		if opts.NewCommitIDs[i] == git.EmptySHA {
+			results = append(results, private.HockProcReceiveRefResult{
+				OriginalRef: opts.RefFullNames[i],
+				OldOID:      opts.OldCommitIDs[i],
+				NewOID:      opts.NewCommitIDs[i],
+				Err:         "Can't delete not exist branch",
+			})
+			continue
+		}
+
+		if !strings.HasPrefix(opts.RefFullNames[i], git.PullRequestPrefix) {
+			results = append(results, private.HockProcReceiveRefResult{
+				IsNotMatched: true,
+				OriginalRef:  opts.RefFullNames[i],
+			})
+			continue
+		}
+
+		baseBranchName := opts.RefFullNames[i][len(git.PullRequestPrefix):]
+		curentTopicBranch := ""
+		if !gitRepo.IsBranchExist(baseBranchName) {
+			// try match refs/for/<target-branch>/<topic-branch>
+			for p, v := range baseBranchName {
+				if v == '/' && gitRepo.IsBranchExist(baseBranchName[:p]) && p != len(baseBranchName)-1 {
+					curentTopicBranch = baseBranchName[p+1:]
+					baseBranchName = baseBranchName[:p]
+					break
+				}
+			}
+		}
+
+		if len(topicBranch) == 0 && len(curentTopicBranch) == 0 {
+			results = append(results, private.HockProcReceiveRefResult{
+				OriginalRef: opts.RefFullNames[i],
+				OldOID:      opts.OldCommitIDs[i],
+				NewOID:      opts.NewCommitIDs[i],
+				Err:         "topic-branch is not set",
+			})
+			continue
+		}
+
+		headBranch := ""
+		userName := strings.ToLower(opts.UserName)
+
+		if len(curentTopicBranch) == 0 {
+			curentTopicBranch = topicBranch
+		}
+
+		if !strings.HasPrefix(curentTopicBranch, userName+"/") {
+			headBranch = userName + "/" + curentTopicBranch
```

Comment:
Maybe we should use userid but not userName because username maybe deleted and recreated. Or we have to delete all these refs files when user is deleted.

Reply:
I think it's not a problem, Because it will't be used to create any git refs. just use to checkout the exist unmerged pull request which has same topic and user name. (see: https://github.com/go-gitea/gitea/pull/14295#issuecomment-869100765)
after pull request closed or merged. This label willn't be used in anywhere.